src/backend/pipeline/cont_combiner.c

```diff
@@ -812,9 +786,18 @@ read_batch(ContQueryCombinerState *state, TupleBufferBatchReader *reader)
 
 		ExecStoreTuple(heap_copytuple(tbs->tuple->heaptup), state->slot, InvalidBuffer, false);
 		tuplestore_puttupleslot(state->batch, state->slot);
+		state->batch_hashes[count] = tbs->tuple->group_hash;
```

Comment:
We can probably avoid allocating this large array for each state by keeping a single global one and re-using it?